# Part I - (Analysis on Prosper Loan Data)
## by (Ampong Boadu)

## Introduction
> The dataset is on loan disbursed to clients between the year 2005 and 2014. The consist of general characteristics of the loan, that is the amount granted, the current status, interest charged, duration of loan captured in months, as well as characteristics of the borrower. Information of the borrowers' were also captured in the angle of the state of the borrower, borrowers' debt to income ratio, thier occupation, information on their income, their credit score among others.



## Preliminary Wrangling

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

Matplotlib is building the font cache; this may take a moment.


In [3]:
# Load data to dataframe

df =  pd.read_csv('prosperLoanData.csv')
df.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380  ...         -133.18                0.0   
1        0.0920       0.0820  ...            0.00                0.0   
2        0.2750       0.2400  ...          -24.20                0.0   
3        0.0974       0.0874  ...         -108.01                0.0   
4        0.2085       0.1985  ...          -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

In [5]:
# Limiting dataframe to columns needed for analysis.

df=df[['LoanKey','LoanStatus','Term','LoanOriginationDate','LoanOriginationQuarter','LoanOriginalAmount',
       'MonthlyLoanPayment','BorrowerAPR','BorrowerRate','DebtToIncomeRatio','IsBorrowerHomeowner','ClosedDate',
       'ListingCategory (numeric)','IncomeRange','ProsperRating (Alpha)','ProsperScore','EmploymentStatus']]

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   LoanKey                    113937 non-null  object 
 1   LoanStatus                 113937 non-null  object 
 2   Term                       113937 non-null  int64  
 3   LoanOriginationDate        113937 non-null  object 
 4   LoanOriginationQuarter     113937 non-null  object 
 5   LoanOriginalAmount         113937 non-null  int64  
 6   MonthlyLoanPayment         113937 non-null  float64
 7   BorrowerAPR                113912 non-null  float64
 8   BorrowerRate               113937 non-null  float64
 9   DebtToIncomeRatio          105383 non-null  float64
 10  IsBorrowerHomeowner        113937 non-null  bool   
 11  ClosedDate                 55089 non-null   object 
 12  ListingCategory (numeric)  113937 non-null  int64  
 13  IncomeRange                11

In [6]:
# Renaming columns to simplier column names

df.rename(columns = {
    'LoanStatus': 'status',
    'Term': 'duration',
    'LoanOriginationDate': 'origination_date',
    'ClosedDate': 'closed_date',
    'LoanOriginalAmount': 'amount',
    'MonthlyLoanPayment': 'monthly_repayment',
    'BorrowerAPR': 'borrower_APR',
    'BorrowerRate': 'borrower_rate',
    'DebtToIncomeRatio': 'debt_income_ratio',
    'IsBorrowerHomeowner': 'home_owner',
    'ListingCategory (numeric)': 'purpose',
    'IncomeRange': 'income_range',
    'ProsperRating (Alpha)': 'prosper_rating',
    'ProsperScore': 'prosper_score',
    'EmploymentStatus': 'employment_status'
}, inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 17 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   LoanKey                 113937 non-null  object 
 1   status                  113937 non-null  object 
 2   duration                113937 non-null  int64  
 3   origination_date        113937 non-null  object 
 4   LoanOriginationQuarter  113937 non-null  object 
 5   amount                  113937 non-null  int64  
 6   monthly_repayment       113937 non-null  float64
 7   borrower_APR            113912 non-null  float64
 8   borrower_rate           113937 non-null  float64
 9   debt_income_ratio       105383 non-null  float64
 10  home_owner              113937 non-null  bool   
 11  closed_date             55089 non-null   object 
 12  purpose                 113937 non-null  int64  
 13  income_range            113937 non-null  object 
 14  prosper_rating      

In [7]:
# Getting year of loan
df['origination_year'] = df.LoanOriginationQuarter.str[3:]
df = df.drop(['LoanOriginationQuarter'], axis = 1)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 17 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   LoanKey            113937 non-null  object 
 1   status             113937 non-null  object 
 2   duration           113937 non-null  int64  
 3   origination_date   113937 non-null  object 
 4   amount             113937 non-null  int64  
 5   monthly_repayment  113937 non-null  float64
 6   borrower_APR       113912 non-null  float64
 7   borrower_rate      113937 non-null  float64
 8   debt_income_ratio  105383 non-null  float64
 9   home_owner         113937 non-null  bool   
 10  closed_date        55089 non-null   object 
 11  purpose            113937 non-null  int64  
 12  income_range       113937 non-null  object 
 13  prosper_rating     84853 non-null   object 
 14  prosper_score      84853 non-null   float64
 15  employment_status  111682 non-null  object 
 16  or

In [8]:
# Checking stats of numerical values in dataframe

df.describe()

duration        amount  monthly_repayment   borrower_APR  \
count  113937.000000  113937.00000      113937.000000  113912.000000   
mean       40.830248    8337.01385         272.475783       0.218828   
std        10.436212    6245.80058         192.697812       0.080364   
min        12.000000    1000.00000           0.000000       0.006530   
25%        36.000000    4000.00000         131.620000       0.156290   
50%        36.000000    6500.00000         217.740000       0.209760   
75%        36.000000   12000.00000         371.580000       0.283810   
max        60.000000   35000.00000        2251.510000       0.512290   

       borrower_rate  debt_income_ratio        purpose  prosper_score  
count  113937.000000      105383.000000  113937.000000   84853.000000  
mean        0.192764           0.275947       2.774209       5.950067  
std         0.074818           0.551759       3.996797       2.376501  
min         0.000000           0.000000       0.000000       1.000000  
25%         0.134000           0.140000       1.000000       4.000000  
50%         0.184000           0.220000       1.000000       6.000000  
75%         0.250000           0.320000       3.000000       8.000000  
max         0.497500          10.010000      20.000000      11.000000

In [9]:
# Dropping null values from Prosper Ratings Columns. The mean is high may influence the results.

df = df[df['prosper_rating'].notna()]

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 84853 entries, 1 to 113936
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   LoanKey            84853 non-null  object 
 1   status             84853 non-null  object 
 2   duration           84853 non-null  int64  
 3   origination_date   84853 non-null  object 
 4   amount             84853 non-null  int64  
 5   monthly_repayment  84853 non-null  float64
 6   borrower_APR       84853 non-null  float64
 7   borrower_rate      84853 non-null  float64
 8   debt_income_ratio  77557 non-null  float64
 9   home_owner         84853 non-null  bool   
 10  closed_date        26005 non-null  object 
 11  purpose            84853 non-null  int64  
 12  income_range       84853 non-null  object 
 13  prosper_rating     84853 non-null  object 
 14  prosper_score      84853 non-null  float64
 15  employment_status  84853 non-null  object 
 16  origination_year   84

In [10]:
# Filling null values in Dept to Income Ratio column with mean value from the column.

df['debt_income_ratio'].fillna((df['debt_income_ratio'].mean()), inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 84853 entries, 1 to 113936
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   LoanKey            84853 non-null  object 
 1   status             84853 non-null  object 
 2   duration           84853 non-null  int64  
 3   origination_date   84853 non-null  object 
 4   amount             84853 non-null  int64  
 5   monthly_repayment  84853 non-null  float64
 6   borrower_APR       84853 non-null  float64
 7   borrower_rate      84853 non-null  float64
 8   debt_income_ratio  84853 non-null  float64
 9   home_owner         84853 non-null  bool   
 10  closed_date        26005 non-null  object 
 11  purpose            84853 non-null  int64  
 12  income_range       84853 non-null  object 
 13  prosper_rating     84853 non-null  object 
 14  prosper_score      84853 non-null  float64
 15  employment_status  84853 non-null  object 
 16  origination_year   84

In [11]:
# Formatting loan purpose

purpose = {
    0: 'Not Available',
    1: 'Debt Consolidation',
    2: 'Home Improvement',
    3: 'Business', 
    4: 'Personal Loan',
    5: 'Student Use',
    6: 'Auto',
    7: 'Other',
    8: 'Baby&Adoption',
    9: 'Boat', 
    10: 'Cosmetic Procedure', 
    11: 'Engagement Ring',
    12: 'Green Loans',
    13: 'Household Expenses',
    14: 'Large Purchases', 
    15: 'Medical/Dental', 
    16: 'Motorcycle', 
    17: 'RV',
    18: 'Taxes',
    19: 'Vacation',
    20: 'Wedding Loans'
}

# Function to return pourpose in String than Value
def purpose_string(num):
    return purpose[num]

In [12]:
# Applying function with lambda
df['purpose'] = df['purpose'].apply(lambda a : purpose_string(a))
df['purpose'].value_counts()

Debt Consolidation    53180
Other                  9218
Home Improvement       6801
Business               5298
Auto                   2237
Household Expenses     1996
Medical/Dental         1522
Taxes                   885
Large Purchases         876
Wedding Loans           771
Vacation                768
Motorcycle              304
Student Use             274
Engagement Ring         217
Baby&Adoption           199
Cosmetic Procedure       91
Boat                     85
Green Loans              59
RV                       52
Not Available            20
Name: purpose, dtype: int64

In [13]:
# Dropping rows with loan purpose not available from dataframe

df = df[df.purpose != 'Not Available']
df['purpose'].value_counts()

Debt Consolidation    53180
Other                  9218
Home Improvement       6801
Business               5298
Auto                   2237
Household Expenses     1996
Medical/Dental         1522
Taxes                   885
Large Purchases         876
Wedding Loans           771
Vacation                768
Motorcycle              304
Student Use             274
Engagement Ring         217
Baby&Adoption           199
Cosmetic Procedure       91
Boat                     85
Green Loans              59
RV                       52
Name: purpose, dtype: int64

In [14]:
df.employment_status.value_counts()

Employed         67294
Full-time         7926
Self-employed     4538
Other             3805
Not employed       648
Retired            367
Part-time          255
Name: employment_status, dtype: int64

In [15]:
print(df.income_range.value_counts())
print(df.prosper_rating.value_counts())

$50,000-74,999    25621
$25,000-49,999    24167
$100,000+         15201
$75,000-99,999    14497
$1-24,999          4654
Not employed        648
$0                   45
Name: income_range, dtype: int64
C     18343
B     15580
A     14550
D     14268
E      9789
HR     6932
AA     5371
Name: prosper_rating, dtype: int64


In [16]:
# Cleaning income range for categorical data
df = df[df['income_range'] != 'Not employed']
df = df[df['income_range'] != '$0']

df.income_range.value_counts()

$50,000-74,999    25621
$25,000-49,999    24167
$100,000+         15201
$75,000-99,999    14497
$1-24,999          4654
Name: income_range, dtype: int64

In [17]:
# Dropping duplicates if they exist
df = df.drop_duplicates(subset = ['LoanKey'])

# Checking if dataset contains duplicates
df[df.duplicated()]

Empty DataFrame
Columns: [LoanKey, status, duration, origination_date, amount, monthly_repayment, borrower_APR, borrower_rate, debt_income_ratio, home_owner, closed_date, purpose, income_range, prosper_rating, prosper_score, employment_status, origination_year]
Index: []

In [18]:
# Checking shape of dataframe
df.shape

(83269, 17)

In [19]:
# Saving clean data to csv file for Explanatory Slides.
df.to_csv('prosper_loans_clean.csv')

In [20]:
# convert income range and ratings into ordered categorical types

ordinal_var_dict = {'income_range': ['$1-24,999','$25,000-49,999','$50,000-74,999','$75,000-99,999','$100,000+'],
                    'prosper_rating': ['HR', 'E', 'D', 'C', 'B', 'A', 'AA'],
                    'status': ['Current', 'Completed', 'Past Due (1-15 days)','Past Due (16-30 days)',
                               'Past Due (31-60 days)','Past Due (61-90 days)', 'Past Due (91-120 days)',
                               'Past Due (>120 days)', 'Chargedoff', 'Defaulted']}

for var in ordinal_var_dict:
    ordered_var = pd.api.types.CategoricalDtype(ordered = True,
                                                categories = ordinal_var_dict[var])
    df[var] = df[var].astype(ordered_var)

### What is the structure of your dataset?

> The dataset contains 113,937 loans in number originated between the year 2005 to 2014. The original datase is almost clean with 81 columns. The wrangling done was in dimension of limiting the dataset to columns that will be necessary the analysis.

### What is/are the main feature(s) of interest in your dataset?

> The focal point is to determine the features of loans in individual status especially Chargedoff and Defaulted loans.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> Features of the loan such as the amount, repayment, status, date of origination together with features of the borrowers.